Add system env to all configs

nikolauska opened the pull request:
This merge adds possibility to use system environment variables on all configs. Having only username and password makes it harder to change other settings when building released package with distillery.

nikolauska commented:
I just realized it's not this easy, I'll add tests and fix this to properly work

kdisneur commented on lib/bamboo/adapters/smtp_adapter.ex:
```diff
@@ -286,6 +286,9 @@ defmodule Bamboo.SMTPAdapter do
   end
 
+  defp to_gen_smtp_server_config({:server, {:system, var}}, config) do
+    [{:relay, System.get_env(var)} | config]
```
While you're at changing the code, what do you think of re-calling the function using the interpolated value? It could make it less error-prone if we don't duplicate the configuration keys.

Example (always easier 😸 )
```elixir
defp to_gen_smtp_server_config({:server, {:system, var}}, config) do
  to_gen_smtp_server_config({:server, System.get_env(var)}, config)
end
```

nikolauska replied:
it could be easier to make it generic like this
```elixir
defp to_gen_smtp_server_config({conf, {:system, var}}, config) do
  to_gen_smtp_server_config({conf, System.get_env(var)}, config)
end
```
and then leave handling different types of configs in specific config for example tls requires atom and System.get_env returns string (this is actually why I realized it wasn't this easy)

nikolauska commented:
I have now reworked settings system and added more validation to make sure values are always correct type. I've also added test for all configs variables to validate that system values are correctly transformed from string to proper format.

tgautier commented:
Seems good to me.

Could you just rebase this to one commit before we accept your PR?

Thanks!

nikolauska commented:
It's now rebased to one commit
